### PR TITLE
✨ Source S3: Add IAM Role Authentication

### PR DESCRIPTION
**Pull request by tolik0**
## What
Enhance the S3 connector source to support IAM role-based authentication, allowing users to utilize IAM roles for more granular control over permissions and to eliminate the need for managing static access keys. Implements: https://github.com/airbytehq/airbyte/issues/31865

## How
[Boto3 class `RefreshableCredentials`](https://github.com/boto/botocore/blob/develop/botocore/credentials.py#L342) is used to refresh credentials automatically during the sync.

### CODE REVIEW

File: airbyte-integrations/connectors/source-s3/source_s3/v4/stream_reader.py [hunk 1]
```diff
@@ -52,14 +58,54 @@ def s3_client(self) -> BaseClient:
...
+                self._s3_client = boto3.client(
+                    "s3",
+                    aws_access_key_id=self.config.aws_access_key_id,
+                    aws_secret_access_key=self.config.aws_secret_access_key,
+                    **client_kv_args,
+                )
+
         return self._s3_client
 
+    def _get_iam_s3_client(self, client_kv_args: dict) -> BaseClient:
```
Comment (bgroff) [inline airbyte-integrations/connectors/source-s3/source_s3/v4/stream_reader.py:74]: This is a nice solution. Can you link any documentation you used for this? I found this article that gave a nice description: https://medium.com/@li.chastina/auto-refresh-aws-tokens-using-iam-role-and-boto3-afd3c52fd8c7

File: docs/integrations/sources/s3.md [hunk 2]
```diff
@@ -45,10 +47,63 @@ At this time, object-level permissions alone are not sufficient to successfully
...
+2. Choose the appropriate trust entity and attach the policy you created. 
+3. Set up a trust relationship for the role. For example for **AWS account** trusted entity use default AWS account on your instance (it will be used to assume role). To use **External ID** set it to environment variables as `export AWS_EXTERNAL_ID="{your-external-id}"`. Edit the trust relationship policy to reflect this:
+```
+{
+    "Version": "2012-10-17",
+    "Statement": [
+        {
+            "Effect": "Allow",
+            "Principal": {
+                "AWS": "arn:aws:iam::{your-aws-account-id}:user/{your-username}"
```
Comment (bgroff) [inline docs/integrations/sources/s3.md:63]: We this is generally supported in cloud, this should probably say: arn:aws:iam::094410056844:user/delegated_access_user for the cloud version only.

File: docs/integrations/sources/s3.md [hunk 3]
```diff
@@ -45,10 +47,63 @@ At this time, object-level permissions alone are not sufficient to successfully
...
+2. Choose the **AWS account** trusted entity type. 
+3. Set up a trust relationship for the role. This allows the Airbyte instance's AWS account to assume this role. You will also need to specify an external ID, which is a secret key that the trusting service (Airbyte) and the trusted role (the role you're creating) both know. This ID is used to prevent the "confused deputy" problem. You can find External ID in **AWS Role ARN** property title or description. Edit the trust relationship policy to include the external ID:
+```
+{
+    "Version": "2012-10-17",
+    "Statement": [
+        {
+            "Effect": "Allow",
+            "Principal": {
+                "AWS": "arn:aws:iam::{airbyte-aws-account-id}:user/{your-username}"
```
Comment (bgroff) [inline docs/integrations/sources/s3.md:86]: This should say: arn:aws:iam::094410056844:user/delegated_access_user

**Review by bgroff — APPROVED** [review]

File: airbyte-integrations/connectors/source-s3/source_s3/v4/stream_reader.py (hunk 1 above)
Comment (tolik0) [reply airbyte-integrations/connectors/source-s3/source_s3/v4/stream_reader.py:74]: I also used the same Medium article but didn't find any additional official documentation to link except for the source code, so I will not add any links.

File: airbyte-integrations/connectors/source-s3/integration_tests/spec.json [hunk 4]
```diff
@@ -365,6 +365,13 @@
         "order": 3,
         "type": "string"
       },
+      "role_arn": {
+        "title": "AWS Role ARN (External ID is '8eee93517501b43a')",
+        "description": "Specifies the Amazon Resource Name (ARN) of an IAM role that you want to use to perform operations requested using this profile. Set External ID as '8eee93517501b43a'.",
+        "airbyte_secret": true,
```
Comment (bgroff) [inline airbyte-integrations/connectors/source-s3/integration_tests/spec.json:371]: Spoke about this privately, but I am in favor of not setting this to secret. This value is not sensitive and I think it is helpful to be able to see it. Very open to push back here though.

Comment (tolik0) [reply airbyte-integrations/connectors/source-s3/integration_tests/spec.json]: I agree with you, as role arn cannot be used outside the users' Airbyte workspace, it should not be secret. I fixed it in the latest commit.

File: airbyte-integrations/connectors/source-s3/source_s3/v4/stream_reader.py (hunk 1 above)
Comment (lazebnyi) [reply airbyte-integrations/connectors/source-s3/source_s3/v4/stream_reader.py:74]: Can you add docstring for this method.

**Review by lazebnyi — APPROVED** [review]
Nice job! 

Need add small description to docstring.

File: airbyte-integrations/connectors/source-s3/source_s3/v4/stream_reader.py (hunk 1 above)
Comment (tolik0) [reply airbyte-integrations/connectors/source-s3/source_s3/v4/stream_reader.py:73]: Added docstring

**Pull request merged**